bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncAdmin.java

```diff
@@ -63,51 +78,148 @@
   private BigtableInstanceName bigtableInstanceName;
+  private final TableAdapter2x tableAdapter2x;
 
   public BigtableAsyncAdmin(BigtableAsyncConnection asyncConnection) throws IOException {
     LOG.debug("Creating BigtableAsyncAdmin");
     this.options = asyncConnection.getOptions();
     this.bigtableTableAdminClient = asyncConnection.getSession().getTableAdminClient();
     this.disabledTables = asyncConnection.getDisabledTables();
     this.bigtableInstanceName = options.getInstanceName();
+    this.tableAdapter2x = new TableAdapter2x(options, new ColumnDescriptorAdapter());
   }
 
   @Override
   public CompletableFuture<Void> createTable(TableDescriptor desc, Optional<byte[][]> splitKeys) {
-    throw new UnsupportedOperationException("createTable"); // TODO
+    if (desc.getTableName() == null) {
+      return FutureUtils.failedFuture(new IllegalArgumentException("TableName cannot be null"));
+    }
+
+    return CompletableFuture.supplyAsync(() -> {
+      CreateTableRequest.Builder builder = CreateTableRequest.newBuilder();
+      builder.setParent(bigtableInstanceName.toString());
+      builder.setTableId(desc.getTableName().getQualifierAsString());
+      builder.setTable(tableAdapter2x.adapt(desc));
+      if (splitKeys.isPresent()) {
+        for (byte[] splitKey : splitKeys.get()) {
+          builder
+              .addInitialSplits(Split.newBuilder().setKey(ByteString.copyFrom(splitKey)).build());
+        }
+      }
+      return builder;
+    }).thenCompose(
+        r -> FutureUtils.toCompletableFuture(bigtableTableAdminClient.createTableAsync(r.build())))
+        .thenAccept(r -> {
+        });
   }
 
   @Override
   public CompletableFuture<Void> createTable(TableDescriptor desc, byte[] startKey, byte[] endKey,
       int numRegions) {
-    throw new UnsupportedOperationException("createTable"); // TODO
+
+    return CompletableFuture.supplyAsync(() -> {
+      Optional<byte[][]> splitKeys = Optional.empty();
+      if (numRegions < 3) {
+        throw new IllegalArgumentException("Must create at least three regions");
+      } else if (Bytes.compareTo(startKey, endKey) >= 0) {
+        throw new IllegalArgumentException("Start key must be smaller than end key");
+      }
+      if (numRegions == 3) {
+        splitKeys = Optional.ofNullable(new byte[][] {startKey, endKey});
+      } else {
+        splitKeys = Optional.ofNullable(Bytes.split(startKey, endKey, numRegions - 3));
+        if (!splitKeys.isPresent() || splitKeys.get().length != numRegions - 1) {
+          throw new IllegalArgumentException("Unable to split key range into enough regions");
+        }
+      }
+      return splitKeys;
+    }).thenCompose(skeys -> createTable(desc, skeys));
   }
 
   @Override
   public CompletableFuture<Void> disableTable(TableName tableName) {
-    throw new UnsupportedOperationException("deleteTable"); // TODO
+    CompletableFuture<Void> cf = new CompletableFuture<>();
+    tableExists(tableName).thenAccept(exists -> {
```

Comment:
why not `thenCompose`?

Reply:
or maybe just return null?

Reply:
This is one of the complexity with the CF framework :) thenCompose is for cases where a CF is propagated down the chain into a another CF. I believe thenAccept (or theRun) variations are the correct forms here which take no params and return nothing.